pkg/cache/sql/db/client.go

```diff
@@ -0,0 +1,321 @@
+and a way to reset the database.
+*/
+package db
+
+import (
+	"bytes"
+	"context"
+	"database/sql"
+	"encoding/gob"
+	"os"
+	"reflect"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/rancher/lasso/pkg/cache/sql/db/transaction"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"modernc.org/sqlite"
+	sqlite3 "modernc.org/sqlite/lib"
+)
+
+const (
+	// InformerObjectCacheDBPath is where SQLite's object database file will be stored relative to process running lasso
+	InformerObjectCacheDBPath = "informer_object_cache.db"
+	// OnDiskInformerIndexedFieldDBPath is where SQLite's indexed fields database file will be stored if, env var value
+	//found at EncryptAllEnvVar is "true".
+	OnDiskInformerIndexedFieldDBPath = "informer_object_fields.db"
+)
+
+// Client is a database client that provides encrypting, decrypting, and database resetting.
+type Client struct {
+	conn      Connection
+	connLock  sync.RWMutex
+	encryptor Encryptor
+	decryptor Decryptor
+}
+
+// Connection represents a connection pool
+type Connection interface {
+	Begin() (*sql.Tx, error)
+	Exec(query string, args ...any) (sql.Result, error)
+	Prepare(query string) (*sql.Stmt, error)
+	Close() error
+}
+
+// Closable Closes an underlying connection and returns an error on failure
+type Closable interface {
+	Close() error
+}
+
+// Rows represents sql rows. It exposes method to navigate the rows, read their outputs, and close them.
+type Rows interface {
+	Next() bool
+	Err() error
+	Close() error
+	Scan(dest ...any) error
+}
+
+// TXClient represents a sql transaction. The TXClient must manage rollbacks as rollback functionality is not exposed.
+type TXClient interface {
+	StmtExec(stmt transaction.Stmt, args ...any) error
+	Exec(stmt string, args ...any) error
+	Commit() error
+	Stmt(stmt *sql.Stmt) transaction.Stmt
+	Cancel() error
+}
+
+// Encryptor encrypts a slice of bytes. It is expected to use data-encryption-key (DEK) which it used to encrypt the slice.
+// It should use a key-encryption-key (KEK) to then encrypt the DEK. The encrypted data, nonce used to encrypt the data,
+// encrypted dek, nonce used to decrypt the dek should be returned, in that order. On failure an error should by returned.
+type Encryptor interface {
+	Encrypt([]byte) ([]byte, []byte, []byte, []byte, error)
+}
+
+// Decryptor decrypts encrypted data given: some encrypted data, nonce used to encrypt the data, an encrypted
+// data-encryption-key (DEK), nonce used to encrypt the DEK. The KEK is not provided. Decryptor is expected to use its
+// receiver to decrypt the DEK using its KEK.
+type Decryptor interface {
+	Decrypt([]byte, []byte, []byte, []byte) ([]byte, error)
+}
+
+var backoffRetry = wait.Backoff{Duration: 50 * time.Millisecond, Factor: 2, Steps: 10}
+
+// NewClient returns a Client. If the given connection is nil then a default one will be created.
+func NewClient(c Connection, encryptor Encryptor, decrypter Decryptor) (*Client, error) {
```

Comment:
This might be deliberate, but I wince because it's spelled `decryptor`. The diff in case is fine

Reply:
Yep, meant to get them all the same my bad.